packages/grid/src/vaadin-grid.js

```diff
@@ -293,8 +293,8 @@ import { StylingMixin } from './vaadin-grid-styling-mixin.js';
  */
 class Grid extends ElementMixin(
   ThemableMixin(
-    DataProviderMixin(
-      ArrayDataProviderMixin(
+    ArrayDataProviderMixin(
+      DataProviderMixin(
```

Comment:
Both `DataProviderMixin` and `ArrayDataProviderMixin` have an observer that is triggered  on data provider callback change. This change ensures `DataProviderMixin`'s observer is always triggered before to ensure the controller has updated reference to the data provider callback. Otherwise, `ArrayDataProviderMixin` will try to load data using the previous data provider here:

https://github.com/vaadin/web-components/blob/9f27a48ff72d2fe62a0d2de01d6c0cd9ae837372/packages/grid/src/vaadin-grid-array-data-provider-mixin.js#L62-L66